test/connector/tcp/mssql/test

```diff
@@ -1,17 +1,133 @@
+      return
+  fi
+
+  echo ""
+  echo "FAIL:"
+  echo ""
+  echo "The string:"
+  echo ""
+  echo "    ${string}"
+  echo ""
+  echo "Did not contain expected substring:"
+  echo ""
+  echo "    ${substring}"
+  echo ""
+
+  exit 1
+}
+
+query_sql_service() {
+  # query_sql_service <server> <port> <user> <password>
+  server=$1
+  port=$2
+  user=$3
+  pw=$4
+
+  docker_cmd="docker-compose run --rm --no-deps"
+  sql_cmd=(sqlcmd -S "$server,$port" -U "$user" -P "$pw" -Q "SELECT 1+1")
+
+  # NOTE: We need to 2>&1 because we need to assert on sqlcmd's error messages
+  # and we need to "|| true" because we don't want the script to exit when this
+  # happens
+  $docker_cmd test "${sql_cmd[@]}" 2>&1 || true
+}
+
+# Automatically detect if we're devmode based on the existence
+# of the secretless-dev container.  We assume that you started
+# your workflow using `./dev` if you are developing, and this
+# command will use the secretless-dev container.
 if [[ ! -z $(docker-compose ps -q secretless-dev) ]]; then
-  export SECRETLESS_HOST=secretless-dev
+  secretless_host="secretless-dev"
 fi
 
-docker-compose run \
-  -e SECRETLESS_HOST="$SECRETLESS_HOST" \
-  --rm \
-  --no-deps \
-  test bash -c "go test -v ./test/connector/tcp/mssql"
+echo "Waiting for '$secretless_host' service to start"
+
+# single quotes are intentional:
+# shellcheck disable=SC2016
+docker-compose run --rm --no-deps test bash -c '
+counter=0
+while ! wget --quiet --output-document - http://'$secretless_host':5335/ready > /dev/null 2>&1; do
+    if expr $counter = 5 > /dev/null; then
+      echo ""
+      echo "Timed out waiting for Secretless"
+      exit 1
+    fi
+    counter=$counter+1
+    >&2 printf ". "
+    sleep 1
+done
+'
+
+echo "'$secretless_host' service is up - continuing "
+echo ""
+echo "Running tests"
+echo ""
+
+if $benchmark; then
+  # TODO: Run these tests in a loop and calculate average metrics
+  # to eliminate jitter.
+  echo "Benchmarking: Query mssql service directly"
+  time good_resp=$(query_sql_service $db_host_tls $db_port $db_user $db_password)
+  assert_contains "$good_resp" "1 rows affected"
+
+  echo "Benchmarking: Query mssql via secretless"
+  time good_resp=$(query_sql_service $secretless_host $secretless_port "NONEXISTENT_USER" "BOGUS_PASSWORD")
+  assert_contains "$good_resp" "1 rows affected"
+
+else
+  echo "Test: Query mssql service directly with correct password succeeds"
+  good_resp=$(query_sql_service $db_host_tls $db_port $db_user $db_password)
+  assert_contains "$good_resp" "1 rows affected"
+
+  echo "Test: Query mssql service directly with incorrect password fails"
+  bad_resp=$(query_sql_service $db_host_tls $db_port "BAD_USER" "BAD_PASSWORD")
+  assert_contains "$bad_resp" "Login failed for user"
+
+  echo "Test: Query mssql via secretless with incorrect password succeeds"
+  good_resp=$(query_sql_service $secretless_host $secretless_port "NONEXISTENT_USER" "BOGUS_PASSWORD")
+  assert_contains "$good_resp" "1 rows affected"
+fi
```

Comment:
Might be worth raising in the main channel, but even though it's cool you were able to add this in a simple way, I think we're mixing concerns by putting the benchmark code here.  I think this should be purely the IT itself.

We already have another tool for benchmarking, so this may be redundant, but it we do want to keep it just a quick and simple benchmark for this code, I think we should move it into its own `benchmark` file that would be run manually and separate from the IT.

Reply:
Yep, if this is redundant, I'll remove it. The only (limited) value to adding a separate `benchmark` script here would be that we'd have something really quick to check that we're not somehow adding obscene amounts of latency with the insertion of secretless into the mix... but if this is being checked regularly in IT, there's not much value gained.

Reply:
Well the `juxtaposer` benchmarks I believe are only run before releases (ofc you _can_ run them whenever you want but normally people don't), so there's some value to moving it into a little benchmark script.

Reply:
I think I'll leave this out for now... Since there would be a lot of shared code between `test` and `benchmark`, it would take a bit of refactoring, e.g creating a test_common file with shared functions, constants, etc.